ocaml/xapi/xapi_xenops.ml

```diff
@@ -613,6 +613,19 @@ module MD = struct
       (List.combine (Range.to_list (Range.make 0 (List.length devs))) devs)
 
   let get_target_pci_address ~__context vgpu =
+    (* Reallocate a vgpu to during vm checkpoint *)
+    let vm = vgpu.Db_actions.vGPU_VM in
+    let implementation = Db.VGPU_type.get_implementation ~__context~self:vgpu.Db_actions.vGPU_type in
+    let vgpu =
+      match Db.VM.get_power_state ~__context ~self:vm, implementation with
+      | `Suspended, `nvidia ->
+        let host = Helpers.get_localhost ~__context in
+        let record = Db.VM.get_record ~__context ~self:vm in
+        Helpers.with_global_lock(fun () -> Vgpuops.create_vgpus ~__context host (vm, record) true);
```

Comment:
This does not look like the right place to call this. It is normally `Message_forwarding.allocate_vm_to_host` that does this. Why is that not happening here? Could you explain the sequence of operations?

Reply:
Thanks. @robhoes 
1. Yes, normally the function `forward_to_suitable_host`  which include `allocate_vm_to_host` does this VM start-`Message_forwarding.start` or  Resume-`Message_forwarding.resume`; However, in vm's checkpoint, it will call `Xapi_xenops.resume` directly at [here](https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_vm_snapshot.ml#L227), so it won't get a valid vGPU through Message_forwarding.
Thus, before this fix, it will raise an exn in [here](https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_xenops.ml#L622) because of there is no vGPU in VM.resume.

2. For the main sequence of operations:
Xapi_vm_snapshot.checkpoint 
-> Xapi_xenops.suspend (it will remove the vGPU allocated in function`Xapi_vm_lifecycle.force_state_reset`) 
-> Xapi_vm_clone.clone (checkpoint operation) 
-> Xapi_xenops.resume

Reply:
Plus, I run more vGPU migration jobs with this fix. 2327822 and 2327823.
As it was merged to feature branch at first, we are developing more test cases base on feature branch. Thanks. @robhoes

Reply:
OK, I see what you mean. However, while it may work in some cases, fixing it this way has a couple of problems:
1. Resource allocation for VMs is important and complicated, and best done from a single place in the code. It is already quite hard to see how this works right now, and adding such calls to Xapi_xenops would make it worse.
2. The job of Xapi_xenops functions like `of_nvidia_vgpu` is to just get information from the DB and turn this into xenopsd config. This patch mixes in resourcing here, which does not fit.
3. The `with_global_lock` function must only be called on the master, usually from Message_forwarding (it's just a mutex), and the same is true for resource allocation functions like `create_vgpus`. Now this code runs on the host where the VM is running, which means that the locking does not work, which can lead to concurrency problems.

Taking a step back, it seems to me that for a VM checkpoint, while internally the VM goes through a suspend/resume cycle, the VM should ideally keep its resources (such as vGPUs), rather than release and reacquire them.

Reply:
Thanks, Rob. You are right, it's not good to mix the resource reservation in xenopsd anymore.
I had refined the code to get all the `record` of the vgpus and then reset the field of `resident_on` if have.
Please help to review again on the latest change. Thanks!